### PR TITLE
Lazily produce semantic models in source generators

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This saves 91% of the time spent during incremental analysis when that analysis only needs syntax.  Turns out getting a semantic model is still somewhat expensive.  So if we don't need it, best not to get it.

### CODE REVIEW

**Review by chsienki — APPROVED** [review]

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+        [IterationCleanup]
+        public void IterationCleanup()
+        {
+            _workspace.Dispose();
+            _workspace = null;
+        }
+
+        [Benchmark]
+        public async Task RunGeneratorOld()
```
Comment (cston) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:110]: >RunGeneratorOld [](http://example.com/codeflow?start=26&length=15)

Should this be named "RunGenerator" instead?

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                    {
+                        var node = (ClassDeclarationSyntax)ctx.Node;
+                        /*
+                        var semanticModel = ctx.SemanticModel;
+                        var classSymbol = semanticModel.GetDeclaredSymbol(node);
+                        foreach (var attr in classSymbol.GetAttributes())
+                        {
+                            attr.AttributeClass.ToDisplayString();
+                        }
+                        */
```
Comment (cston) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:131]: Can we remove the commented out code?

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            _workspace = null;
+        }
+
+        [Benchmark]
+        public async Task RunGeneratorOld()
+        {
+            var generator = (new PipelineCallbackGenerator(ctx =>
+            {
+                Console.WriteLine("Registering");
+#if true
```
Comment (cston) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:115]: Can this be removed?

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                        return node;
+                    });
+#else
+                var input = ctx.ForAttributeWithMetadataName<ClassDeclarationSyntax>("System.Text.Json.Serialization.JsonSerializableAttribute");
+#endif
+                ctx.RegisterSourceOutput(input, (spc, node) => { });
+            })).AsSourceGenerator();
+
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(
+               new ISourceGenerator[] { generator }, parseOptions: CSharpParseOptions.Default);//, driverOptions: new GeneratorDriverOptions(IncrementalGeneratorOutputKind.Source, trackIncrementalGeneratorSteps: false));
```
Comment (cston) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:141]: >//, driverOptions: new GeneratorDriverOptions(IncrementalGeneratorOutputKind.Source, trackIncrementalGeneratorSteps: false)); [](http://example.com/codeflow?start=95&length=125)

Can this be removed?

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:110]: This is just personal benchmark code.  I view it as a playground, vs a distinct section that must follow particular rules.  :)   i can make changes, but it's just not a critical thing from my perspective.

Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:110]:
```suggestion
        public async Task RunGenerator()
```

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:115]: this just helps easily making changes and checking info.

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:131]:
```suggestion
```

Comment (CyrusNajmabadi) [reply src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs]: done.

Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs]:
```suggestion
```

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
```
Comment (CyrusNajmabadi) [inline src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs:133]:
```suggestion
               new ISourceGenerator[] { generator }, parseOptions: CSharpParseOptions.Default);
```

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs]: done.

File: src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Tools/IdeCoreBenchmarks/IncrementalSourceGeneratorBenchmarks.cs]: done.

**Review by cston — APPROVED** [review]

**Comment by 333fred** [comment]
> Turns out getting a semantic model is still somewhat expensive.

Would you be able to elaborate on what is expensive about this? I don't expect it to be, so the assertion is concerning to me.

**Comment by CyrusNajmabadi** [comment]
Gave trace to fred to look at to see if semantic model construct can be made faster.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi: I admit I'm surprised a semantic model is expensive too, so agreeing with @tmat that this seems surprising. Especially since GetSemanticModel doesn't take a cancellation token. Is there a follow up bug to track whatever might need changes there?

**Comment by CyrusNajmabadi** [comment]
I know of know followup issue here.  I've given the traces to compiler if they would like to do anything here.

**Pull request merged**